File: greatest_value_tree_row.py
# // Time Complexity : O(n) 
# // Space Complexity : O(height of the tree) for DFS, O(n) for BFS 
# // Did this code successfully run on Leetcode : yes
# // Any problem you faced while coding this : none

# // Your code here along with comments explaining your approach 

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.res = []
        
    def dfs(self,root,level):
        if not root: return 
        
        if len(self.res) == level: 
            self.res.append(root.val)
            
        self.res[level] = max(self.res[level],root.val)
            
        self.dfs(root.left,level+1)
        self.dfs(root.right,level+1)
        
        
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        if not root: return []
        
        # DFS
        # 
        self.dfs(root,0)
        return self.res
    
    
#         # BFS 
#         queue = deque()
#         res = []
        
#         queue.append(root)
        
#         while queue:
#             size = len(queue)
#             max_level = float('-inf')
#             while size:
#                 top = queue.popleft()
#                 max_level = max(max_level,top.val)
#                 if top.left: queue.append(top.left)
#                 if top.right: queue.append(top.right)
#                 size -= 1
#             res.append(max_level)
#         return res

File: Greatest-Value-Tree-Row.java
// Time Complexity : O(n) where n is length of queue
// Space Complexity : O(h) where h is height of tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

import java.util.*;

//  Definition for a binary tree node.

 
class Solution {

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    
    List<Integer> result;
    Queue<TreeNode> q;
    
    public List<Integer> largestValues(TreeNode root) {
        
        result = new ArrayList<>();
        q = new LinkedList<>();
        
        if (root == null) {
            return result;
        }
        
        q.add(root);
        
        while (!q.isEmpty()) {
            
            int max = Integer.MIN_VALUE;
            int s = q.size();
            
            for (int i = 0; i < s; i++) {
                
                TreeNode temp = q.remove();
                
                // comparing current val with max
                if (temp.val > max) {
                    max = temp.val;
                }
                
                // adding left child
                if (temp.left != null) {
                    q.add(temp.left);
                }
                
                // adding right child
                if (temp.right != null) {
                    q.add(temp.right);
                }
            }
            result.add(max);
        }
        return result;
    }
}
